src/plugins/workspace/public/components/data_source_association/association_data_source_modal.tsx

```diff
@@ -230,27 +253,56 @@ export const AssociationDataSourceModalContent = ({
 
   useEffect(() => {
     setIsLoading(true);
+    setIsRelatedConnectionsLoaded(false);
     getDataSourcesList(savedObjects.client, ['*'])
-      .then((dataSourcesList) => fetchDataSourceConnections(dataSourcesList, http, notifications))
-      .then((connections) => {
-        setAllConnections(connections);
+      .then((dataSourcesList) => {
+        return getOpenSearchAndDataConnections(dataSourcesList, notifications).then(
+          ({ openSearchConnections, dataConnections }) => {
+            setAllConnections([...openSearchConnections, ...dataConnections]);
+            return { openSearchConnections, dataConnections, dataSourcesList };
+          }
+        );
+      })
+      .then(({ openSearchConnections, dataConnections, dataSourcesList }) => {
+        fetchDirectQueryConnections(dataSourcesList, http, notifications).then(
+          (directQueryConnections) => {
+            const updatedOpenSearchConnections = updateFullFillRelatedConnections(
```

Comment:
looks like too much promise, can we use await to avoid callback hell to make it more clear?